src/agent/rustjail/src/cgroups/fs/mod.rs

```diff
@@ -27,1120 +43,848 @@ macro_rules! sl {
-pub const BLKIO_SECTORS: &'static str = "blkio.sectors_recursive";
-pub const BLKIO_IO_SERVICE_BYTES: &'static str = "blkio.io_service_bytes_recursive";
-pub const BLKIO_IO_SERVICED: &'static str = "blkio.io_serviced_recursive";
-pub const BLKIO_IO_QUEUED: &'static str = "blkio.io_queued_recursive";
-pub const BLKIO_IO_SERVICE_TIME: &'static str = "blkio.io_service_time_recursive";
-pub const BLKIO_IO_WAIT_TIME: &'static str = "blkio.io_wait_time_recursive";
-pub const BLKIO_IO_MERGED: &'static str = "blkio.io_merged_recursive";
-pub const BLKIO_TIME: &'static str = "blkio.time_recursive";
-pub const BLKIO_THROTTLE_IO_SERVICE_BYTES: &'static str = "blkio.throttle.io_service_bytes";
-pub const BLKIO_THROTTLE_IO_SERVICED: &'static str = "blkio.throttle.io_serviced";
+        let res = &mut cgroups::Resources::default();
 
-lazy_static! {
-    pub static ref CLOCK_TICKS: f64 = {
-        let n = unsafe { libc::sysconf(libc::_SC_CLK_TCK) };
+        // cpuset
+        if r.cpu.is_some() {
+            info!(sl!(), "cgroup manager set cpu !!!");
+            let cpu = r.cpu.as_ref().unwrap();
 
-        n as f64
-    };
-}
+            let cpuset_controller: &CpuSetController = cg.controller_of().unwrap();
 
-pub fn init_static() {
-    lazy_static::initialize(&DEFAULT_ALLOWED_DEVICES);
-    lazy_static::initialize(&BMAP);
-    lazy_static::initialize(&DMAP);
-    lazy_static::initialize(&BABBRS);
-    lazy_static::initialize(&DABBRS);
-    lazy_static::initialize(&HUGEPAGESIZES);
-    lazy_static::initialize(&CLOCK_TICKS);
-}
+            if !cpu.cpus.is_empty() {
+                cpuset_controller.set_cpus(&cpu.cpus);
+            }
 
-fn write_file<T>(dir: &str, file: &str, v: T) -> Result<()>
-where
-    T: ToString,
-{
-    let p = format!("{}/{}", dir, file);
-    fs::write(p.as_str(), v.to_string().as_bytes())
-        .context(format!("couldn't write to file: {:?}", p))?;
+            if !cpu.mems.is_empty() {
+                // res.cpu.mems = cpu.mems.clone();
+                cpuset_controller.set_mems(&cpu.mems);
+            }
 
-    Ok(())
-}
+            let cpu_controller: &CpuController = cg.controller_of().unwrap();
 
-fn read_file(dir: &str, file: &str) -> Result<String> {
-    let p = format!("{}/{}", dir, file);
-    let ret = fs::read_to_string(p.as_str())?;
-    Ok(ret)
-}
+            let mut shares = cpu.shares.unwrap_or(0);
+            if shares != 0 {
+                if cg.v2() {
+                    shares = convert_shares_to_v2_value(shares);
+                }
+                cpu_controller.set_shares(shares);
+            }
 
-fn copy_parent(dir: &str, file: &str) -> Result<()> {
-    let parent = if let Some(index) = dir.rfind('/') {
-        &dir[..index]
-    } else {
-        return Err(anyhow!("cannot copy file from parent"));
-    };
+            let quota = cpu.quota.unwrap_or(0) as u64;
+            let period = cpu.period.unwrap_or(0);
+            cpu_controller.set_cfs_quota_and_period(quota, period);
```

Comment:
Why do you set shares only if not zero, but quota and period unconditionally?

A value of 0 is unlikely to do any good. If you want to set it to unlimited, it should be `-1`, but the interface of `set_cfs_quota_and_period` has a problem there. It uses `u64`, so you can't easily pass `-1` 😦

Reply:
I opened https://github.com/kata-containers/cgroups-rs/issues/5 as a reminder.